<a href="https://colab.research.google.com/github/PARKHYOUNJUNG/phj_lms23/blob/main/CODE/PHJ_8days.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [3]:
data=pd.read_csv('/content/drive/MyDrive/AI_ADU/sample/insurance.csv')
data.head() 

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [4]:
df=pd.read_csv('/content/drive/MyDrive/AI_ADU/sample/insurance.csv')
df.head() 
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [5]:
df['double bmi']=df['bmi']*2
df

age     sex     bmi  children smoker     region      charges  double bmi
0      19  female  27.900         0    yes  southwest  16884.92400       55.80
1      18    male  33.770         1     no  southeast   1725.55230       67.54
2      28    male  33.000         3     no  southeast   4449.46200       66.00
3      33    male  22.705         0     no  northwest  21984.47061       45.41
4      32    male  28.880         0     no  northwest   3866.85520       57.76
...   ...     ...     ...       ...    ...        ...          ...         ...
1333   50    male  30.970         3     no  northwest  10600.54830       61.94
1334   18  female  31.920         0     no  northeast   2205.98080       63.84
1335   18  female  36.850         0     no  southeast   1629.83350       73.70
1336   21  female  25.800         0     no  southwest   2007.94500       51.60
1337   61  female  29.070         0    yes  northwest  29141.36030       58.14

[1338 rows x 8 columns]

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [6]:
df['debt']=0
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     0  
2          66.00     0  
3          45.41     0  
4          57.76     0  
...          ...   ...  
1333       61.94     0  
1334       63.84     0  
1335       73.70     0  
1336       51.60     0  
1337       58.14     0  

[1338 rows x 9 columns]

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [7]:
df['debt']=df.index
# df['debt']=range(len(df))
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41     3  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]


## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [8]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double bmi        67.54
debt                  1
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [9]:
df.drop(['double bmi'], axis=1, inplace=True)

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [11]:
df[['age','sex']]


age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [12]:
df['age']=df['age']+1
df

age     sex     bmi  children smoker     region      charges  debt
0      20  female  27.900         0    yes  southwest  16884.92400     0
1      19    male  33.770         1     no  southeast   1725.55230     1
2      29    male  33.000         3     no  southeast   4449.46200     2
3      34    male  22.705         0     no  northwest  21984.47061     3
4      33    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   51    male  30.970         3     no  northwest  10600.54830  1333
1334   19  female  31.920         0     no  northeast   2205.98080  1334
1335   19  female  36.850         0     no  southeast   1629.83350  1335
1336   22  female  25.800         0     no  southwest   2007.94500  1336
1337   62  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [14]:
df['charges']=round(df['charges'],2)
# df['charges'].round(2)
df.head()

age     sex     bmi  children smoker     region   charges  debt
0   20  female  27.900         0    yes  southwest  16884.92     0
1   19    male  33.770         1     no  southeast   1725.55     1
2   29    male  33.000         3     no  southeast   4449.46     2
3   34    male  22.705         0     no  northwest  21984.47     3
4   33    male  28.880         0     no  northwest   3866.86     4

In [24]:
def sosu(data):
  return format(data,'.2f')

df['charges'].apply(sosu).head()

0    16884.9
1     1725.5
2     4449.5
3    21984.5
4     3866.9
Name: charges, dtype: object

In [26]:
f = lambda x: '%.2f' % x
df['charges'].apply(f).head()

0    16884.92
1     1725.55
2     4449.46
3    21984.47
4     3866.86
Name: charges, dtype: object

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [22]:
df.sort_index(axis=1).head()

age     bmi   charges  children  debt     region     sex smoker
0   20  27.900  16884.92         0     0  southwest  female    yes
1   19  33.770   1725.55         1     1  southeast    male     no
2   29  33.000   4449.46         3     2  southeast    male     no
3   34  22.705  21984.47         0     3  northwest    male     no
4   33  28.880   3866.86         0     4  northwest    male     no

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [29]:
df.sort_values(by='age').head()

age     sex    bmi  children smoker     region   charges  debt
1248   19  female  39.82         0     no  southeast   1633.96  1248
482    19  female  31.35         0     no  southeast   1622.19   482
492    19  female  25.08         0     no  northeast   2196.47   492
525    19  female  33.88         0     no  southeast  11482.63   525
529    19    male  25.46         0     no  northeast   1708.00   529

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [32]:
df['age'].mean()
# df['age'].sum() / len(df)

40.20702541106129

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

In [33]:
df.head()

age     sex     bmi  children smoker     region   charges  debt
0   20  female  27.900         0    yes  southwest  16884.92     0
1   19    male  33.770         1     no  southeast   1725.55     1
2   29    male  33.000         3     no  southeast   4449.46     2
3   34    male  22.705         0     no  northwest  21984.47     3
4   33    male  28.880         0     no  northwest   3866.86     4

In [34]:
df.corr()

age       bmi  children   charges      debt
age       1.000000  0.109272  0.042469  0.299008 -0.031467
bmi       0.109272  1.000000  0.012759  0.198341 -0.036169
children  0.042469  0.012759  1.000000  0.067998  0.025221
charges   0.299008  0.198341  0.067998  1.000000 -0.003373
debt     -0.031467 -0.036169  0.025221 -0.003373  1.000000

In [35]:
df.cov()

age           bmi    children       charges           debt
age         197.401387      9.362337    0.719303  5.087480e+04    -170.829095
bmi           9.362337     37.187884    0.093795  1.464730e+04     -85.224706
children      0.719303      0.093795    1.453213  9.926743e+02      11.747569
charges   50874.801787  14647.304865  992.674264  1.466524e+08  -15781.315299
debt       -170.829095    -85.224706   11.747569 -1.578132e+04  149298.500000

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


In [36]:
df.head()

age     sex     bmi  children smoker     region   charges  debt
0   20  female  27.900         0    yes  southwest  16884.92     0
1   19    male  33.770         1     no  southeast   1725.55     1
2   29    male  33.000         3     no  southeast   4449.46     2
3   34    male  22.705         0     no  northwest  21984.47     3
4   33    male  28.880         0     no  northwest   3866.86     4

In [37]:
df['sex'].value_counts()
# len(df[df['sex'] == 'female'])
# pd.value_counts(df['sex'])
# pd.value_counts(df.sex)

male      676
female    662
Name: sex, dtype: int64

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

In [40]:
df[df['children'] == 3]

age     sex     bmi  children smoker     region   charges  debt
2      29    male  33.000         3     no  southeast   4449.46     2
7      38  female  27.740         3     no  northwest   7281.51     7
25     60  female  27.720         3     no  southeast  14001.13    25
36     63  female  32.965         3     no  northwest  15612.19    36
54     41  female  28.690         3     no  northwest   8059.68    54
...   ...     ...     ...       ...    ...        ...       ...   ...
1301   63    male  30.875         3    yes  northwest  46718.16  1301
1314   31  female  23.655         3    yes  northwest  18765.88  1314
1320   32    male  31.065         3     no  northwest   5425.02  1320
1332   53  female  44.700         3     no  southwest  11411.68  1332
1333   51    male  30.970         3     no  northwest  10600.55  1333

[157 rows x 8 columns]

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

In [43]:
df[df['children'] >= 5]

age     sex     bmi  children smoker     region   charges  debt
32     20  female  28.600         5     no  southwest   4687.80    32
71     32    male  28.500         5     no  northeast   6799.46    71
166    21  female  37.000         5     no  southwest   4830.63   166
413    26    male  23.900         5     no  southwest   5080.10   413
425    46    male  24.310         5     no  southeast   9788.87   425
438    53  female  46.750         5     no  southeast  12592.53   438
568    50  female  31.900         5     no  southwest  11552.90   568
640    34    male  42.400         5     no  southwest   6666.24   640
877    34    male  33.440         5     no  southeast   6653.79   877
932    47    male  25.800         5     no  southwest  10096.97   932
937    40  female  24.225         5     no  northwest   8965.80   937
969    40  female  34.320         5     no  southeast   8596.83   969
984    21    male  30.115         5     no  northeast   4915.06   984
1085   40  female  18.300         5    yes  southwest  19023.26  1085
1116   42    male  29.640         5     no  northeast   9222.40  1116
1130   40  female  23.870         5     no  southeast   8582.30  1130
1245   29    male  24.300         5     no  southwest   5615.37  1245
1272   44    male  25.520         5     no  southeast  14478.33  1272

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

In [44]:
df[df['bmi'] < 17]

age     sex     bmi  children smoker     region  charges  debt
172    19    male  15.960         0     no  northeast  1694.80   172
428    22  female  16.815         1     no  northeast  3167.46   428
1226   39    male  16.815         2     no  northeast  6640.54  1226

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

In [49]:
df['charges'].groupby(df['sex']).mean()
# df.groupby('sex')['charges'].mean()

sex
female    12569.578716
male      13956.751331
Name: charges, dtype: float64

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

In [50]:
df['charges'].groupby(df['smoker']).mean()

smoker
no      8434.268299
yes    32050.231898
Name: charges, dtype: float64

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

In [58]:
df.groupby(['region'])['charges'].mean()
# df['charges'].groupby(df['region']).mean()

region
northeast    13406.384599
northwest    12417.575169
southeast    14735.411484
southwest    12346.937508
Name: charges, dtype: float64

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

In [59]:
df.groupby(['smoker','region']).mean()
# df[['age','sex']]

smoker  region   
no      northeast     9165.531829
        northwest     8556.463558
        southeast     8032.216264
        southwest     8019.284569
yes     northeast    29673.536269
        northwest    30192.002759
        southeast    34844.997143
        southwest    32269.063966
Name: charges, dtype: float64

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

In [ ]:
df.groupby(['smoker','region'])['charges'].mean()
# df[['age','sex']]

In [60]:
df.head()

age     sex     bmi  children smoker     region   charges  debt
0   20  female  27.900         0    yes  southwest  16884.92     0
1   19    male  33.770         1     no  southeast   1725.55     1
2   29    male  33.000         3     no  southeast   4449.46     2
3   34    male  22.705         0     no  northwest  21984.47     3
4   33    male  28.880         0     no  northwest   3866.86     4

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

In [61]:
df.groupby(['smoker','region']).count()

age  sex  bmi  children  charges  debt
smoker region                                           
no     northeast  257  257  257       257      257   257
       northwest  267  267  267       267      267   267
       southeast  273  273  273       273      273   273
       southwest  267  267  267       267      267   267
yes    northeast   67   67   67        67       67    67
       northwest   58   58   58        58       58    58
       southeast   91   91   91        91       91    91
       southwest   58   58   58        58       58    58

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>

In [69]:
smgu = lambda sm : ('비흡연자') if(sm=='no') else ('흡연자')
df['group_val']=df['smoker'].apply(smgu)
df.head()
# weight2 = lambda man, height : (height - 100) * 0.9 if(man) else (height - 100)

age     sex     bmi  children smoker     region   charges  debt group_val
0   20  female  27.900         0    yes  southwest  16884.92     0       흡연자
1   19    male  33.770         1     no  southeast   1725.55     1      비흡연자
2   29    male  33.000         3     no  southeast   4449.46     2      비흡연자
3   34    male  22.705         0     no  northwest  21984.47     3      비흡연자
4   33    male  28.880         0     no  northwest   3866.86     4      비흡연자

In [70]:
df.drop(['group_val'], axis=1, inplace=True)

In [71]:
df2 = pd.DataFrame({'group_val' : ['흡연자', '비흡연자']}, index =['yes', 'no'])

In [72]:
df2

group_val
yes       흡연자
no       비흡연자

In [73]:
pd.merge(df, df2, left_on='smoker', right_index = True, how = 'outer' )

age     sex    bmi  children smoker     region   charges  debt group_val
0      20  female  27.90         0    yes  southwest  16884.92     0       흡연자
11     63  female  26.29         0    yes  southeast  27808.73    11       흡연자
14     28    male  42.13         0    yes  southeast  39611.76    14       흡연자
19     31    male  35.30         0    yes  southwest  36837.47    19       흡연자
23     35  female  31.92         1    yes  northeast  37701.88    23       흡연자
...   ...     ...    ...       ...    ...        ...       ...   ...       ...
1332   53  female  44.70         3     no  southwest  11411.68  1332      비흡연자
1333   51    male  30.97         3     no  northwest  10600.55  1333      비흡연자
1334   19  female  31.92         0     no  northeast   2205.98  1334      비흡연자
1335   19  female  36.85         0     no  southeast   1629.83  1335      비흡연자
1336   22  female  25.80         0     no  southwest   2007.94  1336      비흡연자

[1338 rows x 9 columns]

In [77]:
df['group_val']=df['smoker'].replace(['yes','no'],['흡연자','비흡연자'])
df.head()

age     sex     bmi  children smoker     region   charges  debt group_val
0   20  female  27.900         0    yes  southwest  16884.92     0       흡연자
1   19    male  33.770         1     no  southeast   1725.55     1      비흡연자
2   29    male  33.000         3     no  southeast   4449.46     2      비흡연자
3   34    male  22.705         0     no  northwest  21984.47     3      비흡연자
4   33    male  28.880         0     no  northwest   3866.86     4      비흡연자